src/core/Table/hooks/useColumnDragAndDrop.tsx

```diff
@@ -0,0 +1,140 @@
+  isEnabled: boolean,
+) => (hooks: Hooks<T>) => {
+  hooks.getDragAndDropProps = [defaultGetDragAndDropProps(isEnabled)];
+  hooks.useInstance.push(useInstance);
+};
+
+const defaultGetDragAndDropProps = <T extends Record<string, unknown>>(
+  isEnabled: boolean,
+) => (
+  props: TableKeyedProps,
+  {
+    instance,
+    header,
+  }: {
+    instance: TableInstance;
+    header: HeaderGroup;
+  },
+) => {
+  if (!isEnabled || header.disableDragging) {
+    return props;
+  }
+
+  const onDragStart = (event: React.DragEvent<HTMLDivElement>) =>
+    event.dataTransfer.setData('text', header.id);
+
+  const setOnDragColumnStyle = (
+    event: React.DragEvent<HTMLDivElement>,
+    position?: 'left' | 'right',
+  ) => {
+    const columnElement = event.currentTarget as HTMLElement;
+    columnElement.classList.remove('iui-reorder-column-right');
+    columnElement.classList.remove('iui-reorder-column-left');
+    if (position === 'left') {
+      columnElement.classList.add('iui-reorder-column-left');
+    } else if (position === 'right') {
+      columnElement.classList.add('iui-reorder-column-right');
+    }
+  };
+
+  const reorderColumns = (
+    tableColumns: IdType<T>[],
+    srcIndex: number,
+    dstIndex: number,
+  ) => {
+    const newTableColumns = [...tableColumns];
+    const [removed] = newTableColumns.splice(srcIndex, 1);
+    newTableColumns.splice(
+      // When dragging column to the right, we need to decrease the index
+      // because we removed column before and indexes shifted to the left.
+      dstIndex > srcIndex ? dstIndex - 1 : dstIndex,
+      0,
+      removed,
+    );
+    return newTableColumns;
+  };
+
+  const onDragOver = (event: React.DragEvent<HTMLDivElement>) => {
+    event.preventDefault();
+    const columnElement = event.target as HTMLElement;
+    const middlePoint =
+      columnElement.offsetLeft + columnElement.offsetWidth / 2;
+    setOnDragColumnStyle(event, event.clientX > middlePoint ? 'right' : 'left');
+  };
+
+  const onDragLeave = (event: React.DragEvent<HTMLDivElement>) => {
+    setOnDragColumnStyle(event);
+  };
+
+  const onDrop = (event: React.DragEvent<HTMLDivElement>) => {
+    event.preventDefault();
+    setOnDragColumnStyle(event);
+
+    const srcColumnId = event.dataTransfer.getData('text');
+    const columnIds = instance.flatHeaders.map((x) => x.id);
+    const srcIndex = columnIds.findIndex((x) => x === srcColumnId);
+    const dstIndex = columnIds.findIndex((x) => x === header.id);
+
+    if (srcIndex === dstIndex) {
+      return;
+    }
+
+    const columnElement = event.target as HTMLElement;
+    const middlePoint =
+      columnElement.offsetLeft + columnElement.offsetWidth / 2;
+
+    instance.setColumnOrder(
+      reorderColumns(
+        columnIds,
+        srcIndex,
+        // When dropped on the right side of the column, need to increase the index by 1
+        event.clientX > middlePoint ? dstIndex + 1 : dstIndex,
```

Comment:
I'm not sure about this middle point logic. If you look at windows explorer for example, it reorders as soon as it enters the destination column.

Reply:
I like that it it appears either on left or right of the destination column. And it shows where its gonna land.

Reply:
Here's a gif demonstrating the problem with the current middle point approach. You have to move the mouse too far into the next column. The blue feedback makes it even more confusing why nothing is happening

![dragging](https://user-images.githubusercontent.com/9084735/158381371-904b9100-092d-48b6-9e6d-e603b3813076.gif)

Compared with Windows Explorer, where it performs the reorder as soon as the mouse enters the next column.
![dragging2](https://user-images.githubusercontent.com/9084735/158381875-ea01e862-fc4e-497d-8c20-f435def9706e.gif)

Reply:
Here is how macOS handles it for comparison:
![column-reorder](https://user-images.githubusercontent.com/849817/158393076-0553d5ab-5222-4fdc-8963-291942e5f717.gif)